heron/healthmgr/src/java/com/twitter/heron/healthmgr/common/PhysicalPlanProvider.java

```diff
@@ -0,0 +1,104 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.healthmgr.common;
+
+import java.net.MalformedURLException;
+import java.util.logging.Logger;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+
+import com.microsoft.dhalion.events.EventHandler;
+import com.microsoft.dhalion.events.EventManager;
+
+import com.twitter.heron.healthmgr.common.HealthManagerEvents.ContainerRestart;
+import com.twitter.heron.healthmgr.common.HealthManagerEvents.TopologyUpdate;
+import com.twitter.heron.proto.system.PhysicalPlans.PhysicalPlan;
+import com.twitter.heron.proto.system.PhysicalPlans.StMgr;
+import com.twitter.heron.spi.statemgr.SchedulerStateManagerAdaptor;
+
+import static com.twitter.heron.healthmgr.HealthManager.CONF_TOPOLOGY_NAME;
+
+/**
+ * A topology's physical plan may get updated after initial deployment. This provider is used to
+ * fetch the latest version from the state manager and provide to any dependent components.
+ */
+public class PhysicalPlanProvider implements Provider<PhysicalPlan> {
+  private static final Logger LOG = Logger.getLogger(PhysicalPlanProvider.class.getName());
+
+  private final SchedulerStateManagerAdaptor stateManagerAdaptor;
+  private final String topologyName;
+
+  private PhysicalPlan physicalPlan;
+
+  @Inject
+  public PhysicalPlanProvider(SchedulerStateManagerAdaptor stateManagerAdaptor,
+      EventManager eventManager, @Named(CONF_TOPOLOGY_NAME) String topologyName) {
+    this.stateManagerAdaptor = stateManagerAdaptor;
+    this.topologyName = topologyName;
+    eventManager.addEventListener(TopologyUpdate.class, new EventHandler<TopologyUpdate>() {
+      /**
+       * Invalidates cached physical plan on receiving topology update notification
+       */
+      @Override
+      public synchronized void onEvent(TopologyUpdate event) {
+        LOG.info("Received topology update event, invalidating cached PhysicalPlan: " + event);
+        physicalPlan = null;
+      }
+    });
+    eventManager.addEventListener(ContainerRestart.class, new EventHandler<ContainerRestart>() {
+      /**
+       * Invalidates cached physical plan on receiving container restart notification
+       */
+      @Override
+      public synchronized void onEvent(ContainerRestart event) {
+        LOG.info("Received conatiner restart event, invalidating cached PhysicalPlan: " + event);
+        physicalPlan = null;
+      }
+    });
+  }
+
+  /**
+   * get the heron-shell url by stmgr id
+   * @param stmgrId the stmgr id in the same container as heron-shell
+   * @return heron-shell url. throw exception if the stmgr id is not found in the phyiscal plan
+   */
+  public String getShellUrl(String stmgrId) throws MalformedURLException {
```

Comment:
This method does not belong to this class. This class should just provide the physical plan. Creation of shell url should live in some utility class?

Also, save as above, why not use the scheduler's restart command as used by the command line?

Reply:
removed